repoaudit/repoaudit/__init__.py

```diff
@@ -60,32 +92,41 @@ def main() -> None:
 
 @main.command()
 @recursive_option
-@click.argument("url")
-@click.option("--dists", help="Comma separated list of distributions.")
+@click.argument("url", required=False)
+@click.option("--dists", required=False, help="Comma separated list of distributions.")
+@click.option(
+    "--apt-source",
+    "-a",
+    required=False,
+    type=click.File("r"),
+    help=(
+        "Supply apt sources.list file. When provided, all entries"
+        "will be parsed for repo urls and their respective dists."
+    )
+)
 @file_option
 @pubkey_option
-def apt(recursive: bool, url: str, dists: str, output: str, pubkeys: str) -> None:
+def apt(recursive: bool, url: str, dists: str, apt_source: str, output: str, pubkeys: str) -> None:
     """Validate an apt repository at URL."""
-    if recursive:
-        urls = get_repo_urls(url)
-    else:
-        urls = [url]
-
-    if dists:
-        dist_set = set(dists.split(","))
-    else:
-        dist_set = None
-
-    errors = RepoErrors()
-
-    with _gpg_cmdline(pubkeys) as gpg:
+    if apt_source:
         try:
-            for repo_url in urls:
-                check_apt_repo(repo_url, dist_set, gpg, errors)
-        except KeyboardInterrupt:
-            pass
-
-    output_result(errors, output)
+            lines = [line.strip() for line in apt_source if line.startswith("deb")]
+            for line in lines:
+                fields = line.split(" ")
+                index: int
+                if (fields[1].startswith("http")):
+                    index = 1
+                elif (fields[2].startswith("http")):
+                    index = 2
+                else:
+                    raise Exception("No URL found")
+                url = fields[index]
+                dists = ','.join(fields[(index + 1):])
```

Comment:
I think a line goes:

```
deb url dist comp1 comp2 comp3
```

In other words, I think there's only one dist per line and the rest are comps.

Reply:
It would be better to use a regular expression here if you're up for the challenge. This should get you started:

```python
import re

src_line = re.compile("(deb|deb-src)( \[.*\])? (?P<url>http.*?) .*")

match = re.fullmatch(src_line, "deb [arch=amd64 signed-by=/usr/share/keyrings/some-repo.gpg] http://some.repo/apt dev main")
match["url"]  # 'http://some.repo/apt'

match = re.fullmatch(src_line, "deb-src http://site.example.com/debian distribution component1 component2 component3")
match["url"]. # 'http://site.example.com/debian'
```

Once you have that down this stuff simply becomes:

```python
for line in lines:
    match = re.fullmatch(src_line, line)
    url = match["url"]
    dist = match["dist"]
```

This is totally optional but I think would make a good intro to regular expressions.

Reply:
More links about regexs in python:

https://docs.python.org/3/library/re.html
https://docs.python.org/3/howto/regex.html
https://developers.google.com/edu/python/regular-expressions